[5.0] Fix fps drop on high polling rate mouse movement

JoeTwizzle opened the pull request:
### Purpose of this PR

* This PR fixes the low performance of raw mouse input on the Win32 backend.
* It rewrites the ProcessEvents function to handle RawInput events separately from other events. Additionally it also avoids processing more than 5 events per ProcessEvents call. This does not cause issues as the events are automatically aggregated by Win32.

Before:
https://github.com/user-attachments/assets/81b49fe1-8190-42ad-8f33-c0c9f67101d1

After:
https://github.com/user-attachments/assets/a392e505-f540-455a-bd41-dc9f683d141f


### Testing status

* Tested on the ``VulkanTestProject`` for the frame-time graph, and in addition tested on the ``OpenTK.Backends.Tests`` project to test the actual mouse values.

### Comments

* I am not 100% confident that this solution is ideal for multi window applications. This is because it relies on the window being focused in order to process raw input events. Maybe it should be changed to being handled by a window hover mechanism?

JoeTwizzle committed:
Fix fps drop on high polling rate mouse movement

JoeTwizzle committed:
Dynamic allocation for raw input buffer

NogginBops commented on src/OpenTK.Platform/Native/Windows/WindowComponent.cs:
```diff
@@ -1143,21 +1108,120 @@ private IntPtr Win32WindowProc(IntPtr hWnd, WM uMsg, UIntPtr wParam, IntPtr lPar
+                else break;
+            } while (true);
+
+            IntPtr currentRawInputPtr = _rawInputBuffer;
```
Would probably be good if this was a typed pointer instead of `IntPtr`.

NogginBops commented on src/OpenTK.Platform/Native/Windows/WindowComponent.cs:
```diff
@@ -1143,21 +1108,120 @@ private IntPtr Win32WindowProc(IntPtr hWnd, WM uMsg, UIntPtr wParam, IntPtr lPar
+
+            for (uint i = 0; i < fetchCount && _rawInputMessagesProcessedThisFrame < MaxRawInputsPerFrame; i++)
+            {
+                Win32.RAWINPUT rawInput = Marshal.PtrToStructure<Win32.RAWINPUT>(currentRawInputPtr);
```
No, need to marshal this struct. `Win32.RAWINPUT` is unmanaged and blittable.

NogginBops commented on src/OpenTK.Platform/Native/Windows/Win32.cs:
```diff
@@ -1471,6 +1471,13 @@ internal struct RAWINPUTDEVICE
 
+        [DllImport("User32.dll", SetLastError = true)]
+        public static extern uint GetRawInputBuffer(
+            [Out, Optional] IntPtr pData,  // PRAWINPUT
```
Follow the format of all the other functions. The actual type of the IntPtr should be between the type and the variable name. For this function it doesn't make sense to use `IntPtr` when `RAWINPUT*` is the more correct signature.

Don't use the `[Out]` and `[Optional]` attribute.

NogginBops reviewed (changes requested):
I do think it's better to read the entire raw input buffer in one go like here and limiting the number of raw input events delivered every frame is a reasonable idea.

It seems like this implementation discards all of the raw input messages that are not dispatched in the frame, this is probably also something we should consider more carefully.

JoeTwizzle committed:
Address feedback pt1

Fix indentation in helper window switch
Change GetRawInputBuffer signature to match the format of other fuctions.
Change RawMouseMoveEventArgs to inherit from EventArgs instead of WindowEventArgs.
Add MaxRawInputsPerFrame to ToolkitOptions.WindowsOptions
Fix memory leak (my bad)
Use NativeMemory instead of AllocCoTaskMem.

JoeTwizzle committed:
Add Settings for Mouse input processing (Windows)

Adds MaxRawMouseMessagesPerFrame and MaxMouseMoveMessagesPerFrame to ToolkitOptions.WindowsOptions.

JoeTwizzle committed:
Rewrite raw input handling + add mouse event limit

NogginBops commented on src/OpenTK.Platform/Native/Windows/WindowComponent.cs:
```diff
@@ -1143,21 +1107,88 @@ private IntPtr Win32WindowProc(IntPtr hWnd, WM uMsg, UIntPtr wParam, IntPtr lPar
+                    if (error != Win32.ERROR_INSUFFICIENT_BUFFER)
+                    {
+                        throw new Win32Exception(error, "GetRawInputBuffer (fetch data) failed.");
+                    }
```
This should not be an error, we should handle this properly. If you want I can rewrite this if you don't want to.

JoeTwizzle committed:
Fix comment spacing

NogginBops committed:
Moved raw input handling to a separate thread in the windows backend.

NogginBops committed:
Add comment

NogginBops committed:
Fixed logging.

NogginBops committed:
Fixed if-statement

NogginBops reviewed (approved):
@JoeTwizzle I've reworked this PR to use the same approach that SDL takes, which is to do all raw input processing on a separate thread and pass the results of that to the main thread. I also handle the `ERROR_INSUFFICIENT_BUFFER` result properly by resizing the buffer and trying again.

NogginBops commented on src/OpenTK.Platform/Native/Windows/WindowComponent.cs:
```diff
@@ -1143,13 +1109,207 @@ private IntPtr Win32WindowProc(IntPtr hWnd, WM uMsg, UIntPtr wParam, IntPtr lPar
             }
         }
 
+        // 512 should be enough space to handle 16k Hz mouse at 30fps without
```
```suggestion
        // 512 should be enough space to handle 16k Hz mouse at 30fps without blocking
```

NogginBops committed:
Fix comment

NogginBops committed:
Add faster raw input queue. Use NEXTRAWINPUTBLOCK to traverse raw input buffer.

NogginBops committed:
Fixed to use AlignedFree for RawInputBuffer to avoid heap corruption crash on exit. Do not throw exception when DPI awareness has already been set for the process. Convert a bunch of Console.WriteLine to use a logger.

NogginBops committed:
Merge branch 'opentk5.0' of github.com:opentk/opentk into PR-1882-high-poll-rate-perf

NogginBops committed:
Merge branch 'opentk5.0' of github.com:opentk/opentk into PR-1882-high-poll-rate-perf

NogginBops committed:
Remove unused import.

NogginBops committed:
Try to limit performance impact of legacy mouse messages with high-DPI mice.
